### ISSUE TITLE
Help with customization

### ISSUE THREAD

**Issue by geraldus** [company-mode/company-mode#484]
Hi @dgutov !

I have a question about proper customization of `company-backends` variable.  Most time I writing Haskell code with `haskell-mode` enabled.  It provides completion-at-point function which added to `completion-at-point-functions` hook.  Most of the time it works fine, however there are several caveats.  Haskell mode's completion function do not works inside comments and strings (except special type of comments — language directives called "pragmas" in Haskell).  Honestly, Haskell mode have two completion functions, one for static mode and other for interactive mode, the former provides simple completions from built-in lists, and the latter uses interactive REPL process to provide accurate completions for module names and identifiers.

From this point I will assume interactive mode use case.  One of drawbacks of interactive completions that it can't provide candidates for local identifiers.  As workaround I override `company-backends` value in Haskell buffers prepending `(company-capf company-dabbrev-code)` grouped backend to default value.  This work very nice, however now I have no completions in comments and strings at all!  I can solve this by adding `company-dabbrev` to my grouped backend, but this makes candidates list outside comments and string a bit «noisy».

I supposed that if first backend fails to return results company will run next backends in `company-backends` list, however it looks like not the case.  Perfectly, I want completions to come from `(company-capf company-dabbrev-code)` when not in comment or string, otherwise from `company-dabbrev` only.  Is it possible to achieve this somehow?

**Comment by dgutov** [comment]
How about

```el
(setq-local company-backends '((company-capf company-dabbrev-code) company-dabbrev)
```

?

**Comment by geraldus** [comment]
Thank you (:
Actually, I have already tested this approach without success, but your comment made me to take closer look at what's happening.  So, I've found minor bug in haskell mode's static completion function.  Rather than return `nil` in comments it returned a list with prefix boundaries positions and nil as a candidate.  Fixing that makes `company-dabbrev` to work in comments.

Thank you again for amazing `company-mode` (:
